lib/src/revset.rs

```diff
@@ -193,6 +193,21 @@ impl error::Error for RevsetParseError {
 pub const GENERATION_RANGE_FULL: Range<u32> = 0..u32::MAX;
 pub const GENERATION_RANGE_EMPTY: Range<u32> = 0..0;
 
+/// Symbol or function to be resolved to `CommitId`s.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum RevsetCommitRef {
+    Symbol(String),
+    // TODO: VisibleHeads
```

Comment:
FYI, we may still want to somehow tell the revset engine which the visible heads are. A scenario we've talked about internally is if the user asks for children or descendants of a hidden commit. Since thats's restricted to the visible heads, it would be useful to be able to tell the user that (part of) their revset expression was empty because of that.

It's also possible to do that at the UI level, at some extra cost. That would probably end up much cleaner. We should be able to pretty cheaply check if any symbols resolved to hidden commits and only do further checking if they did. The UI layer is of course also able to provide better hints. And there may be legitimate reasons to allow evaluation of `descendants(<hidden commit>)` without any kind of warning or error from the library code.

Reply:
> FYI, we may still want to somehow tell the revset engine which the visible heads are. A scenario we've talked about internally is if the user asks for children or descendants of a hidden commit.

Yes, I agree visible heads are a bit special. For this particular scenario, we can teach visible heads through `Children/Descendants { visible_heads: Vec<CommitId>, .. }`, but I don't know if that's a good idea.

As you said, it might be better to deal with that at UI layer, or still in library but engine-agnostic way.